tempesta_fw/sched/tfw_sched_http.c

```diff
@@ -195,276 +243,404 @@ tfw_sched_http_cfg_arg_tbl[_TFW_HTTP_MATCH_F_COUNT] = {
+tfw_cfgop_http_rule(TfwCfgSpec *cs, TfwCfgEntry *e)
 {
 	int r;
-	size_t arg_size;
-	TfwSchedHttpRule *rule;
-	tfw_http_match_op_t op;
-	tfw_http_match_fld_t field;
-	tfw_http_match_arg_t type;
-	TfwVhost *vhost;
-	const char *in_vhost, *in_field, *in_op, *in_arg;
-
-	if ((r = tfw_cfg_check_val_n(e, 4)))
-		return r;
-
-	in_vhost = e->vals[0];
-	in_field = e->vals[1];
-	in_op = e->vals[2];
-	in_arg = e->vals[3];
+	bool invert;
+	TfwHttpMatchRule *rule;
+	const char *in_field, *arg, *action, *action_val;
+	tfw_http_match_op_t op = TFW_HTTP_MATCH_O_WILDCARD;
+	tfw_http_match_fld_t field = TFW_HTTP_MATCH_F_WILDCARD;
+	tfw_http_match_arg_t type = TFW_HTTP_MATCH_A_WILDCARD;
+	TfwCfgRule *cfg_rule = &e->rule;
+	size_t arg_size = 0;
+	TfwHttpChain *chain = NULL;
+	TfwVhost *vhost = NULL;
 
-	vhost = tfw_vhost_lookup(in_vhost);
-	if (!vhost) {
-		TFW_ERR_NL("sched_http: vhost is not found: '%s'\n",
-			   in_vhost);
-		r = -EINVAL;
-		goto err;
+	BUG_ON(!tfw_chain_entry);
+	if ((r = tfw_cfg_check_val_n(e, 0)))
+		return r;
+	if (e->attr_n) {
+		TFW_ERR_NL("Attibutes count must be zero\n");
+		return -EINVAL;
 	}
 
-	r = tfw_cfg_map_enum(tfw_sched_http_cfg_field_enum, in_field, &field);
-	if (r) {
-		TFW_ERR_NL("sched_http: invalid HTTP request field: '%s'\n",
-			   in_field);
-		r = -EINVAL;
-		goto err;
+	invert = cfg_rule->inv;
+	in_field = cfg_rule->fst;
+	arg = cfg_rule->snd;
+	action = cfg_rule->act;
+	action_val = cfg_rule->val;
+	BUG_ON(!action);
+
+	if (tfw_http_rule_default_exist(&tfw_chain_entry->match_list)) {
+		TFW_ERR_NL("sched_http: default HTTP rule must be"
+			   " only one and last; chain '%s'\n",
+			   tfw_chain_entry->name ? : "main" );
+		return -EINVAL;
 	}
 
-	r = tfw_cfg_map_enum(tfw_sched_http_cfg_op_enum, in_op, &op);
-	if (r) {
-		TFW_ERR_NL("sched_http: invalid matching operator: '%s'\n",
-			   in_op);
-		r = -EINVAL;
-		goto err;
+	/* Interpret condition part of the rule. */
+	if (in_field && tfw_http_arg_adjust(&arg, &arg_size, &op)) {
+		r = tfw_cfg_map_enum(tfw_sched_http_cfg_field_enum,
+				     in_field, &field);
+		if (r) {
+			TFW_ERR_NL("sched_http: invalid rule field: '%s'\n",
+				   in_field);
+			return r;
+		}
+		type = tfw_sched_http_cfg_arg_tbl[field];
 	}
-
-	arg_size = strlen(in_arg) + 1;
-	type = tfw_sched_http_cfg_arg_tbl[field];
-
-	rule = tfw_http_match_entry_new(tfw_rules_reconfig,
-					TfwSchedHttpRule, rule, arg_size);
+	rule = tfw_http_rule_new(tfw_chain_entry, type, arg_size);
 	if (!rule) {
-		TFW_ERR_NL("sched_http: can't allocate memory for parsed "
-			   "rule\n");
-		r = -ENOMEM;
-		goto err;
+		TFW_ERR_NL("sched_http: can't allocate memory for rule\n");
+		return -ENOMEM;
+	} else if (type == TFW_HTTP_MATCH_A_STR ||
+		   type == TFW_HTTP_MATCH_A_NUM ||
+		   type == TFW_HTTP_MATCH_A_WILDCARD)
+	{
+		rule->inv = invert;
+		r = tfw_http_rule_init(rule, field, op, type, arg, arg_size - 1);
+		if (r)
+			return r;
+	} else {
+		BUG();
+		// TODO: parsing rules of other types
```

Comment:
Please don't use C++-style for persistent comments.

Why don't you parse the other types?

Reply:
`Addr` and `method` types are not used in TempestaFW.
Corrected.